packages/context/src/app/AppProvider.tsx

```diff
@@ -62,8 +78,57 @@ export type AppProviderProps = {
 
 function AppProvider(props: AppProviderProps) {
     const { app, children } = props;
-
-    return <AppContext.Provider value={app}>{children}</AppContext.Provider>;
+    const apiServices = useApiServices();
+
+    const appWithFunctions: AppWithFunctions = {
+        ...app,
+        /**
+         * Looks up the default page slug for a given object or its nearest type ancestor.
+         *
+         * @param {string} objectId - The ID of the object to start the search from.
+         * @returns {Promise<string | undefined>} The default page slug, or `undefined` if no default page is found.
+         */
+        findDefaultPageSlugFor: useCallback(
+            async (objectId: string) => {
+                let defaultPageId: string | undefined;
+                const objectHierarchy = [objectId];
+                let currentObjectId: string | undefined = objectId;
+                while (currentObjectId !== undefined) {
+                    if (app.defaultPages?.[currentObjectId]) {
+                        defaultPageId = app.defaultPages[currentObjectId];
+                        break;
+                    }
+
+                    let effectiveObject: Obj | undefined;
+                    try {
+                        effectiveObject = await apiServices.get<Obj>(`data/objects/${currentObjectId}/effective`, {
+                            params: { filter: { fields: ['baseObject'] }, sanitizedVersion: true },
+                        });
+                    } catch (error) {
+                        console.error(error);
+                    }
+                    if (effectiveObject?.baseObject?.objectId) {
+                        objectHierarchy.push(effectiveObject?.baseObject?.objectId);
+                    }
+                    currentObjectId = effectiveObject?.baseObject?.objectId ?? undefined;
+                }
+
+                let defaultPage: Page | undefined;
+                if (defaultPageId && typeof defaultPageId === 'string') {
+                    const pageId = defaultPageId.includes('/')
+                        ? defaultPageId.split('/').slice(2).join('/')
+                        : defaultPageId;
+                    defaultPage = await apiServices.get<Page>(
+                        `/webContent/apps/${app.id}/pages/${encodeURIComponent(encodeURIComponent(pageId))}`,
```

Comment:
Have you tested that this works?  Not sure if this will generate `https://...evokeplatform.com/api/webContent/...` or `https://...evokeplatform.com/webContent/...`.  BTW would be nice to have unit tests.

Reply:
Yeah, it works. It ends up with /api on it. We add that to the baseUrl before passing that to the ApiBaseUrl Provider.